### PR TITLE
ci: add cross-rs build

### LINKED ISSUE #951: Ryot crash since 6.9.3
Hi,
until version 6.9.2 everything was working fine for me, but since 6.9.3 I'm getting:

```
[backend] /bin/sh: 1: /usr/local/bin/ryot: Exec format error
[backend] BACKEND_PORT=5000 /usr/local/bin/ryot exited with code 126
--> Sending SIGTERM to other processes..
node:events:492
      throw er; // Unhandled 'error' event
      ^

Error: spawn ps ENOENT
    at ChildProcess._handle.onexit (node:internal/child_process:286:19)
    at onErrorNT (node:internal/child_process:484:16)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
Emitted 'error' event on ChildProcess instance at:
    at ChildProcess._handle.onexit (node:internal/child_process:292:12)
    at onErrorNT (node:internal/child_process:484:16)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'spawn ps',
  path: 'ps',
  spawnargs: [ '-o', 'pid', '--no-headers', '--ppid', 14 ]
}

Node.js v20.10.0
[backend] /bin/sh: 1: /usr/local/bin/ryot: Exec format error
[backend] BACKEND_PORT=5000 /usr/local/bin/ryot exited with code 126
--> Sending SIGTERM to other processes..
node:events:492
      throw er; // Unhandled 'error' event
      ^

Error: spawn ps ENOENT
    at ChildProcess._handle.onexit (node:internal/child_process:286:19)
    at onErrorNT (node:internal/child_process:484:16)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)
Emitted 'error' event on ChildProcess instance at:
    at ChildProcess._handle.onexit (node:internal/child_process:292:12)
    at onErrorNT (node:internal/child_process:484:16)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'spawn ps',
  path: 'ps',
  spawnargs: [ '-o', 'pid', '--no-headers', '--ppid', 14 ]
}

Node.js v20.10.0
```

Maybe it's an issue with the ARM version?

### PR DESCRIPTION
**Pull request by vnghia**
Fixes #951.

### CODE REVIEW

**Comment by IgnisDa** [comment]
Hi! Was this closed by mistake?

**Comment by vnghia** [comment]
@IgnisDa, please test the latest `ghcr.io/vnghia/ryot:test`, feel free to edit the PR and ping me if you need help

**Comment by IgnisDa** [comment]
It does not seem to be working on my macbook.

**Comment by IgnisDa** [comment]
It is working fine on My server which is x86.

File: .github/workflows/main.yaml [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+          locked: false
+
+      - name: Build
+        env:
+          APP_VERSION: ${{ steps.build.outputs.version }}
+          DEFAULT_TMDB_ACCESS_TOKEN: ${{ secrets.DEFAULT_TMDB_ACCESS_TOKEN }}
+          DEFAULT_MAL_CLIENT_ID: ${{ secrets.DEFAULT_MAL_CLIENT_ID }}
+        run: |
+          touch .env
+          ${{ matrix.platform.command }} build --locked --target ${{ matrix.platform.target }} --profile ${{ steps.build.outputs.profile }}
```
Comment (IgnisDa) [inline .github/workflows/main.yaml]: Please build it in release mode once you are done with testing it.

**Review by IgnisDa — COMMENTED** [review]
Additionally, I think the old docker file and release workflow can be removed too.

**Comment by IgnisDa** [comment]
Also, if since you have been so helpful in contributing, I want to give you a pro key for free. Just drop me a DM on the discord if you want. Link is in the readme.

File: .github/workflows/main.yaml (hunk 1 above)
Comment (IgnisDa) [reply .github/workflows/main.yaml]: NVM look like the profile will be automatically selected dependending on the release type.

File: ci/cross/Dockerfile.aarch64 [hunk 2]
```diff
@@ -2,6 +2,7 @@ FROM ubuntu:22.04
 ARG DEBIAN_FRONTEND=noninteractive
 
 RUN apt-get update && apt-get install --assume-yes --no-install-recommends \
+    build-essential \
```
Comment (IgnisDa) [inline ci/cross/Dockerfile.aarch64]: maybe `libssl-dev` too?

**Comment by vnghia** [comment]
I think we are good to go. The problem with `arm64` is actually quite complex. 
- `cross-rs` uses an Ubuntu Docker image which contains all the tools needed for cross-compiling. Setting up them manually is well impossible as you can see my efforts here.
- The Ubuntu version (`20.04`) is quite outdated at this point but they don't have enough time to update it yet. `20.04` only has `libssl1.1` which is linked against the binary while compiling.
- The Debian bookworm we are using as the final image only has `libssl3` and not `libssl1.1`. The final solutions I came up with is to manually install `libssl1.1` from the previous version packages. This is the disvantage of dynamically linking, you have to make sure that the shared library is the same for both compiling and running phase.

>  I want to give you a pro key for free. Just drop me a DM on the discord if you want it.

Thank you very much. Just too busy right now to get my life organized, I will drop you a DM when I have sometime xD.

> Please build it in release mode once you are done with testing it.

Some note about profile and caching. It will build with `dev` for PR and push to main and `release` for tag. Caching only happens when there is a push to main (either a commit or a tag)

**Comment by vnghia** [comment]
you can test it with `ghcr.io/vnghia/ryot:pr-6`

**Comment by IgnisDa** [comment]
I have a habit of making a lot of commits. Don't you think it would be better that the image is not built on every commit on a PR? I think it should run only when the commit message contains `RUN CI` or some other mechanism.

**Comment by vnghia** [comment]
Actually, github already has the mechanism to skip the ci based on the commit message so may be you can use it ? https://github.blog/changelog/2021-02-08-github-actions-skip-pull-request-and-push-workflows-with-skip-ci/

**Comment by IgnisDa** [comment]
These are more like opt out models. I want the images to be built on an OPT IN basis. So when my commit message has a specific substring, only then the worflow should trigger.

Does that make sense?

File: .github/workflows/main.yaml [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+          locked: false
+
+      - name: Build
+        env:
+          APP_VERSION: ${{ steps.build.outputs.version }}
+          DEFAULT_TMDB_ACCESS_TOKEN: ${{ secrets.DEFAULT_TMDB_ACCESS_TOKEN }}
+          DEFAULT_MAL_CLIENT_ID: ${{ secrets.DEFAULT_MAL_CLIENT_ID }}
+          DEFAULT_GOOGLE_BOOKS_API_KEY: ${{ secrets.DEFAULT_GOOGLE_BOOKS_API_KEY }}
+        run: |
+          touch .env
```
Comment (IgnisDa) [inline .github/workflows/main.yaml]: Is this file needed anymore?

File: .github/workflows/main.yaml [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        uses: docker/setup-buildx-action@v3
+
+      - name: Log in to the ghcr container registry
+        uses: docker/login-action@v3
+        with:
+          registry: ${{ env.GHCR_REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Extract metadata for Docker
```
Comment (IgnisDa) [inline .github/workflows/main.yaml]: Could you bring back the old method of getting docker image names? The image needs to be published to ghcr and dockerhub.

**Comment by IgnisDa** [comment]
> The Debian bookworm we are using as the final image only has libssl3 and not libssl1.1. The final solutions I came up with is to manually install libssl1.1 from the previous version packages. This is the disvantage of dynamically linking, you have to make sure that the shared library is the same for both compiling and running phase.

Will it help to get rid of all the manual installations if we use some other image as the base? I can definitely see the current Dockerfile becoming too complex for me to wrap my head around.

File: Dockerfile [hunk 5]
```diff
@@ -21,53 +19,31 @@ COPY --from=frontend-workspace /app/.moon/docker/sources .
...
+ENV TARGETARCH=${TARGETARCH}
 LABEL org.opencontainers.image.source="https://github.com/IgnisDa/ryot"
 ENV FRONTEND_UMAMI_SCRIPT_URL="https://umami.diptesh.me/script.js"
 ENV FRONTEND_UMAMI_WEBSITE_ID="5ecd6915-d542-4fda-aa5f-70f09f04e2e0"
 COPY --from=caddy:2.7.5 /usr/bin/caddy /usr/local/bin/caddy
-RUN apt-get update && apt-get install -y --no-install-recommends curl ca-certificates procps && rm -rf /var/lib/apt/lists/*
+RUN apt-get update && apt-get install -y --no-install-recommends libssl3 ca-certificates procps && rm -rf /var/lib/apt/lists/*
 RUN npm install --global concurrently@8.2.2 && concurrently --version
 RUN useradd -m -u 1001 ryot
+RUN if [ "${TARGETARCH}" = "arm64" ]; then apt-get update && apt-get install -y --no-install-recommends wget && wget http://ftp.debian.org/debian/pool/main/o/openssl/libssl1.1_1.1.1w-0+deb11u1_arm64.deb && dpkg -i libssl1.1_1.1.1w-0+deb11u1_arm64.deb && rm -rf libssl1.1_1.1.1w-0+deb11u1_arm64.deb && apt-get remove -y wget && rm -rf  rm -rf /var/lib/apt/lists/*; fi
```
Comment (vnghia) [inline Dockerfile:39]: @IgnisDa the only manual installation is this line and it is only a workaround until `cross-rs` update their base image. Right now, we have
- `amd64` compiles on github runner which has `libssl3`.
- `arm64` compiles on `cross-rs` base image which has `libssl1.1`.

Finding a base image that has both libraries is impossible so i think we should just stick with this.

File: .github/workflows/main.yaml (hunk 3 above)
Comment (vnghia) [reply .github/workflows/main.yaml]: I am not sure but the CI does not work without that `.env`. I am not sure how the `dotenvy` crate works

File: .github/workflows/main.yaml (hunk 4 above)
Comment (vnghia) [reply .github/workflows/main.yaml]: can you test it ? i dont have a dockerhub account

File: Dockerfile (hunk 5 above)
Comment (IgnisDa) [reply Dockerfile:39]: Cool thanks

File: .github/workflows/main.yaml (hunk 3 above)
Comment (IgnisDa) [reply .github/workflows/main.yaml]: Ah okay, I understand why it happens. The problem with building in debug mode is that pr images will not be able to communicate with TMDb etc.

I think we should always compile in release mode. Especially since now the images are built only when they are manually triggerred.

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_TOKEN }}
+
+      - name: Extract metadata for Docker
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          images: |
+            name=${{ secrets.DOCKER_USERNAME }}/${{ github.event.repository.name }}
+            name=${{ env.GHCR_REGISTRY }}/${{ github.actor }}/${{ github.event.repository.name }}
```
Comment (IgnisDa) [inline .github/workflows/main.yml]: how does `github.actor` work when someone else is making a PR. Will it take `ignisda` or the username of the person who made the PR?

Comment (vnghia) [reply .github/workflows/main.yml]: hmm it seems that it is the one who makes the PR. Let me dig into github action. It is a mess :/

Comment (IgnisDa) [reply .github/workflows/main.yml]: I think we can just go ahead and hardcode `ignisda` there?

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -0,0 +1,198 @@
+name: Main
```
Comment (IgnisDa) [inline .github/workflows/main.yml:1]: the create release and kodi build also seem to be missing. We need to add them back.

**Review by IgnisDa — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yml (hunk 7 above)
Comment (vnghia) [reply .github/workflows/main.yml:1]: How is your release workflow ?

Comment (IgnisDa) [reply .github/workflows/main.yml:1]: The release for the major version is edited with latest details:

https://github.com/IgnisDa/ryot/blob/15a3fe9c35675184e662ea774f6feb71415cbbe8/.github/workflows/release.yml#L26-L63

Kodi artifacts are built and deployed, docs are updated and the demo instance is also redeployed:

https://github.com/IgnisDa/ryot/blob/15a3fe9c35675184e662ea774f6feb71415cbbe8/.github/workflows/release.yml#L137-L195

Comment (IgnisDa) [reply .github/workflows/main.yml:1]: I think this flow might need to be revisited. Does not seem optimal and I have had a few users complain about it not being very predicatable.

File: .github/workflows/main.yml (hunk 6 above)
Comment (vnghia) [reply .github/workflows/main.yml]: I already add a fix in this commit: [`4a6f237` (#939)](https://github.com/IgnisDa/ryot/pull/939/commits/4a6f23734355974460317a2f13e44abcb8cf10e4)

File: .github/workflows/main.yml (hunk 7 above)
Comment (vnghia) [reply .github/workflows/main.yml:1]: One thing I will add is the workflow should not depend on the `create-release` action (at least not the PR workflow) because if the contributors want to test on their local repo, they have to create a release.

Comment (IgnisDa) [reply .github/workflows/main.yml:1]: Yeah I'll make the workflow create a new release with auto generated body for only tagged released and attach the kodi archive as an artifact.

Comment (vnghia) [reply .github/workflows/main.yml:1]: Do you need my help on anything else ?

Comment (IgnisDa) [reply .github/workflows/main.yml:1]: What's the use of building the frontend twice? I think we can get rid of the `build-frontend` step completely and do that part inside the dockerfile.

Comment (IgnisDa) [reply .github/workflows/main.yml:1]: Made the changes, can you take a look and see if it looks fine?

Comment (vnghia) [reply .github/workflows/main.yml:1]: Yeah if it work, it's fine. That is because I dont know how moon works so I just copy from the old dockerfile but I am happy that we dont need to build the frontend twice
